openquake/hazardlib/gsim/bayless_abrahamson_2018.py

```diff
@@ -0,0 +1,519 @@
+#!/usr/bin/env python
+# coding: utf-8
+
+# In[1]:
+
+
+"""
+Module exports :class:`BaylessAbrahamson2018`
+"""
+
+import numpy as np
+from openquake.hazardlib.gsim.base import GMPE, CoeffsTable
+from openquake.hazardlib import const
+from openquake.hazardlib.imt import EAS
+
+
+def _magnitude_scaling(C, ctx):
+    """ Compute the magnitude scaling term """
+    t1 = C['c1']
+    t2 = C['c2'] * (ctx.mag - 6.)
+    tmp = np.log(1.0 + np.exp(C['cn']*(C['cM']-ctx.mag)))
+    t3 = ((C['c2'] - C['c3']) / C['cn']) * tmp
+    return t1 + t2 + t3
+
+
+def _path_scaling(C, ctx):
+    """ Compute path scaling term """
+    tmp = np.maximum(ctx.mag-C['chm'], 0.)
+    t1 = C['c4'] * np.log(ctx.rrup + C['c5'] * np.cosh(C['c6'] * tmp))
+    tmp = (ctx.rrup**2 + 50.**2)**0.5
+    t2 = (-0.5-C['c4'])*np.log(tmp)
+    t3 = C['c7'] * ctx.rrup
+    return t1 + t2 + t3
+
+
+def _normal_fault_effect(C, ctx):
+    """ Compute the correction coefficient for the normal faulting """
+
+    fnm = 0
+    idx = (ctx.rake > -150) & (ctx.rake < -30)
+    fnm = np.zeros_like(ctx.rake)
+    fnm[idx] = 1
+
+    return C['c10'] * fnm
+
+
+def _ztor_scaling(C, ctx):
+    """ Compute top of rupture term """
+    return C['c9'] * np.minimum(ctx.ztor, 20.)
+
+
+def _get_ln_ir_outcrop(self, ctx):
+    """
+    Compute the natural logarithm of peak PGA at rock outcrop, Ir.
+    See eq.10e page 2092.
+    """
+    # Set imt and cofficients. The value of frequency is consistent with
+    # the one used in the matlab script - line 156
+    im = EAS(5.0118725)
+    tc = self.COEFFS[im]
+    # Get mean
+    mean = (_magnitude_scaling(tc, ctx) +
+            _path_scaling(tc, ctx) +
+            _ztor_scaling(tc, ctx) +
+            _normal_fault_effect(tc, ctx))
+    return 1.238 + 0.846 * mean
+
+
+def _linear_site_response(C, ctx):
+    """ Compute the site response term """
+
+    tmp = np.minimum(ctx.vs30, 1000.)
+    fsl = C['c8'] * np.log(tmp / 1000.)
+    return fsl
+
+
+def _soil_depth_scaling(C, ctx):
+    """ Compute the soil depth scaling term """
+    # Set the c11 coefficient - See eq.13b at page 2093
+
+    c11 = np.ones_like(ctx.vs30) * C['c11a']
+    c11[(ctx.vs30 <= 300) & (ctx.vs30 > 200)] = C['c11b']
+    c11[(ctx.vs30 <= 500) & (ctx.vs30 > 300)] = C['c11c']
+    c11[(ctx.vs30 > 500)] = C['c11d']
+    # Compute the Z1ref parameter
+    tmp = (ctx.vs30**4 + 610**4) / (1360**4 + 610**4)
+    z1ref = 1/1000. * np.exp(-7.67/4*np.log(tmp))
+    # Return the fz1 parameter
+    tmp = np.minimum(ctx.z1pt0, np.ones_like(ctx.z1pt0)*2.0) + 0.01
+    return c11 * np.log(tmp / (z1ref + 0.01))
+
+
+def _get_stddevs(C, ctx):
+    """
+    Compute the standard deviations
+    """
+    # Set components of std
+
+    tau = np.zeros_like(ctx.mag)
+    phi_s2s = np.zeros_like(ctx.mag)
+    phi_ss = np.zeros_like(ctx.mag)
+
+    tau[(ctx.mag < 4)] = C['s1']
```

Comment:
You are using a lot of redundant parenthesis. Thet can be removed in most places below.

Reply:
I cleaned this part since the code is not yet vectorised by ruptures.